File: ingestion/nodejs_fallback/src/services/gcsService.js
// =============================================================================
// File: src/services/gcsService.js
// Purpose: Provides a helper function to upload files to Google Cloud Storage
//          using the @google-cloud/storage SDK.
//          Bucket name is read from the GCS_BUCKET_NAME environment variable.
// =============================================================================

// TODO: const { Storage } = require('@google-cloud/storage');
// TODO: const storage = new Storage();

/**
 * Upload a file buffer to GCS.
 *
 * @param {Buffer} fileBuffer - Raw file bytes.
 * @param {string} destinationBlobName - Target path/name inside the GCS bucket.
 * @returns {Promise<string>} gs:// URI of the uploaded object.
 *
 * TODO: Get bucket name from process.env.GCS_BUCKET_NAME.
 * TODO: Create a bucket reference and blob reference.
 * TODO: Save fileBuffer to the blob using blob.save().
 * TODO: Return the gs:// URI.
 */
async function uploadToGcs(fileBuffer, destinationBlobName) {
  // TODO: Implement GCS upload
  throw new Error('uploadToGcs is not yet implemented');
}

module.exports = { uploadToGcs };

File: ingestion/python_api/services/gcs_service.py
# =============================================================================
# Purpose: Provides helper functions for uploading and downloading files from
#          Google Cloud Storage (GCS) using the google-cloud-storage SDK.
#          Bucket name and credentials are read from environment variables.
# =============================================================================

import os
from google.cloud import storage
from dotenv import load_dotenv

load_dotenv()


def _get_bucket():
    bucket_name = os.getenv("GCS_BUCKET_NAME")
    if not bucket_name:
        raise ValueError("GCS_BUCKET_NAME environment variable is not set.")
    client = storage.Client()
    return client.bucket(bucket_name)


def upload_to_gcs(file_bytes: bytes, destination_blob_name: str) -> str:
    """
    Upload raw bytes to a GCS bucket.

    Args:
        file_bytes: Raw bytes of the file to upload.
        destination_blob_name: Target path/name inside the GCS bucket.

    Returns:
        gs:// URI of the uploaded object.
    """
    if not file_bytes:
        raise ValueError("file_bytes cannot be empty.")

    if not destination_blob_name:
        raise ValueError("destination_blob_name is required.")

    bucket = _get_bucket()
    blob = bucket.blob(destination_blob_name)

    # Upload from memory
    blob.upload_from_string(file_bytes)

    return f"gs://{bucket.name}/{destination_blob_name}"


def download_from_gcs(source_blob_name: str) -> bytes:
    """
    Download a file from GCS and return its bytes.

    Args:
        source_blob_name: Path/name of the blob inside the GCS bucket.

    Returns:
        Raw bytes of the downloaded file.
    """
    if not source_blob_name:
        raise ValueError("source_blob_name is required.")

    bucket = _get_bucket()
    blob = bucket.blob(source_blob_name)

    if not blob.exists():
        raise FileNotFoundError(f"Blob '{source_blob_name}' does not exist.")

    return blob.download_as_bytes()